contentcuration/contentcuration/frontend/channelEdit/vuex/contentNode/actions.js

```diff
@@ -42,6 +43,165 @@ export function loadAncestors(context, { id, channel_id }) {
+
+  // Ids of immediate previous/next steps
+  let relatedNodesIds = [
+    ...Object.keys(prerequisite_mapping),
+    ...Object.keys(postrequisite_mapping),
+  ];
+  // + ids of their parent nodes
+  prerequisite_tree_nodes.forEach(node => {
+    if (relatedNodesIds.includes(node.id)) {
+      relatedNodesIds.push(node.parent);
+    }
+  });
+  // remove duplicate ids if any
+  relatedNodesIds = [...new Set(relatedNodesIds)];
+
+  // Make sure that client has all related nodes data available
+  try {
+    await loadContentNodes(context, { ids: relatedNodesIds });
+  } catch (error) {
+    return Promise.reject(error);
+  }
+
+  return Promise.resolve();
+}
+
+/**
+ * Remove a previous step from a target content node.
+ *
+ * @param {String} targetId ID of a target content node.
+ * @param {String} previousStepId ID of a content node to be removed
+ *                                from target's content node previous steps.
+ */
+export function removePreviousStepFromNode(context, { targetId, previousStepId }) {
+  const targetNode = context.state.contentNodesMap[targetId];
+  const targetNodePreviousSteps = targetNode.prerequisite
+    ? targetNode.prerequisite.filter(id => id !== previousStepId)
+    : [];
+
+  context.commit('REMOVE_PREVIOUS_STEP', { targetId, previousStepId });
+  context.commit('UPDATE_CONTENTNODE', {
+    id: targetId,
+    ...{ prerequisite: targetNodePreviousSteps },
+  });
+
+  ContentNode.update(targetId, { prerequisite: targetNodePreviousSteps });
+
+  // (re)load the previous step node to be sure that we have its
+  // `is_prerequisite_of` field up-to-date on client
+  loadContentNode(context, previousStepId);
+}
+
+/**
+ * Remove a next step from a target content node.
+ *
+ * @param {String} targetId ID of a target content node.
+ * @param {String} nextStepId ID of a content node to be removed
+ *                            from target's content node next steps.
+ */
+export function removeNextStepFromNode(context, { targetId, nextStepId }) {
+  removePreviousStepFromNode(context, {
+    targetId: nextStepId,
+    previousStepId: targetId,
+  });
+}
+
+/**
+ * Add a previous step to a target content node.
+ *
+ * @param {String} targetId ID of a target content node.
+ * @param {String} previousStepId ID of a content node to be added
+ *                                to target's content node previous steps.
+ */
+export async function addPreviousStepToNode(context, { targetId, previousStepId }) {
+  const targetNode = context.state.contentNodesMap[targetId];
+  const targetNodePreviousSteps = targetNode.prerequisite || [];
+
+  if (!targetNodePreviousSteps.includes(previousStepId)) {
+    targetNodePreviousSteps.push(previousStepId);
+  }
+
+  await updateContentNode(context, {
+    id: targetId,
+    prerequisite: targetNodePreviousSteps,
+  });
+
+  // (re)load the previous step node to be sure that we have its
+  // `is_prerequisite_of` field up-to-date on client
+  loadContentNode(context, previousStepId);
```

Comment:
Looking over the overall structure, I wonder if it might be best to explicitly load the PrerequisiteRelationship table for direct editing (rather than trying to handle it as a many to many related field on ContentNode) - the only complexity would be doing the extra lookups on the client side, but I think this might be a better way to go.

Thoughts?

Reply:
I like the idea. It would solve a couple of issues I've experienced and it also relates to some ideas I've had abound backend part mostly in relation to things discussed https://github.com/learningequality/studio/pull/1787#discussion_r396116362. Most of the inconsistencies I would like to get rid of is caused by keeping this information duplicate. It would be great to customize backend and sync to our new architecture on this opportunity. Besides other things, I would suggest removing prerequisites data from ContentNode API response if we are going to keep relationships on the frontend.

Though it would be good to talk about this with @jayoshih because when we discussed a couple of sync issues and backend in general, we agreed on using what we have at this point so we can move on because everything's changing fast and also because of time scope for this task. It's also something I would also like to find some time to think it through. Would it make sense as follow-up?

Reply:
> There is some built in caching in the resource layer that might mean if this gets called frequently, the update will not occur as intended. There are two ways to handle this...

Ah had no idea about this, thanks for the info. Originally I made this update on frontend as you suggest in 1. and then I removed it because I assumed that it may be a bit more consistent to let backend do its work and just fetch up-to-date data again. Going with 1. sounds fine to me for now. And later keeping `PrerequisiteRelationship` on the client will solve it even better I guess.

Reply:
Follow-up issue here https://github.com/learningequality/studio/issues/1817